onyo/lib/tests/test_git.py

```diff
@@ -0,0 +1,264 @@
+    `GitRepo.find_root()` must raise `OnyoInvalidRepoError` when called on a
+    Path which is not a git repository.
+    """
+    # try with an existing but non-repo path
+    with pytest.raises(OnyoInvalidRepoError):
+        GitRepo.find_root(tmp_path)
+    # Path that does not exist at all
+    with pytest.raises(OnyoInvalidRepoError):
+        GitRepo.find_root(Path("This/path/does/not/exist"))
+
+
+@pytest.mark.repo_files('a/test/asset_for_test.0',
+                        'a/test/asset_for_test.1')
+def test_GitRepo_restore_staged(repo: OnyoRepo) -> None:
+    """
+    `GitRepo.restore_staged()` must restore all staged files in the repository.
+    If no files are staged, it should not raise an error.
+    """
+    new_git = GitRepo(repo.git.root)
+    test_file = Path('a/test/asset_for_test.0').resolve()
+    test_file2 = Path('a/test/asset_for_test.1').resolve()
+
+    # test that no error is raised, when called on a clean repository
+    new_git.restore_staged()
+
+    # change two files and add them, so they are "staged for commit"
+    test_file.open('w').write('Test: content')
+    test_file2.open('w').write('Test: content')
+    new_git.add([test_file, test_file2])
+    assert test_file in new_git.files_staged
+    assert test_file2 in new_git.files_staged
+
+    # call restore_staged() and verify that the changes are not staged anymore
+    new_git.restore_staged()
+    assert test_file not in new_git.files_staged
+    assert test_file2 not in new_git.files_staged
+
+
+@pytest.mark.repo_files('a/test/asset_for_test.0',
+                        'a/test/asset_for_test.1')
+def test_GitRepo_restore(repo: OnyoRepo) -> None:
+    """
+    `GitRepo.restore()` receives a list of paths and must restore changes for
+    them.
+    """
+    new_git = GitRepo(repo.git.root)
+    test_file = Path('a/test/asset_for_test.0').resolve()
+    test_file2 = Path('a/test/asset_for_test.1').resolve()
+
+    # test that no error is raised, when called on a clean repository
+    new_git.restore([test_file, test_file2])
+
+    # change two files and add them, so they are "staged for commit"
+    test_file.open('w').write('Test: content')
+    test_file2.open('w').write('Test: content')
+    assert test_file in new_git.files_changed
+    assert test_file2 in new_git.files_changed
+
+    # call `GitRepo.restore()` on one of the two modified files, and verify
+    # that the changes for this are not staged anymore, but still for the other
+    new_git.restore(test_file)
+    assert test_file not in new_git.files_changed
+    assert test_file2 in new_git.files_changed
+
+
+@pytest.mark.repo_files('a/test/asset_for_test.0')
+def test_GitRepo_clear_caches(repo: OnyoRepo) -> None:
+    """
+    The function `GitRepo.clear_caches()` must allow to empty the cache of the
+    GitRepo, so that an invalid cache can be re-loaded by a new call of the
+    property.
+    """
+    new_git = GitRepo(repo.git.root)
+    file = Path('a/test/asset_for_test.0').resolve()
```

Comment:
This `resolve` confuses me (may be I'm wrong about how the fixtures work - I hope I'm not). Shouldn't that path be relative to the repo, when you use it with `pytest.mark.repo_files`? Now you treat it as being relative to CWD. That makes no sense. That would change the test depending on where I run the test from.

And if I'm wrong, we'd need to change that. Running the tests should never create things in my CWD!

Reply:
I will change them all to paths from the repo.root.